src/visitable.cpp

```diff
@@ -528,31 +528,31 @@ static long charges_of_internal( const T& self, const itype_id& id )
             if( e->typeId() == id || ( e->is_tool() && e->type->tool->subtype == id ) ) {
                 qty += e->ammo_remaining(); // includes charges from any contained magazine
             }
-            return VisitResponse::SKIP;
+            return qty != limit ? VisitResponse::SKIP : VisitResponse::ABORT;
 
         } else if( e->count_by_charges() ) {
             if( e->typeId() == id ) {
                 qty += e->charges;
             }
             // items counted by charges are not themselves expected to be containers
-            return VisitResponse::SKIP;
+        return qty != limit ? VisitResponse::SKIP : VisitResponse::ABORT;
         }
 
         // recurse through any nested containers
-        return VisitResponse::NEXT;
+        return qty != limit ? VisitResponse::NEXT : VisitResponse::ABORT;
```

Comment:
This should probably be `qty < limit`. `qty` could be increased from a value lesser than `limit` to a value more than `limit`, without ever reaching the `limit` itself.